### PR TITLE
Use foreign VRF name as evpn.transit_vni to implement complex topologies

### PR DESCRIPTION
**Pull request by ipspace**
This is an alternate solution to #502. Instead of turning off uniqueness checks, allow a VRF evpn.transit_vni to refer to another VRF.

### CODE REVIEW

File: netsim/modules/evpn.py [hunk 1]
```diff
@@ -105,6 +105,8 @@ def vrf_transit_vni(topology: Box) -> None:
   for vrf_name,vrf_data in topology.vrfs.items():               # Second pass: set transit VNI values for VRFs with "transit_vni: True"
     if vrf_data is None:                                        # Skip empty VRF definitions
```
Comment (jbemmel) [inline netsim/modules/evpn.py:106]: topology.vrfs get normalized by the vrf module, this case should not happen

File: netsim/modules/evpn.py [hunk 2]
```diff
@@ -116,6 +118,27 @@ def vrf_transit_vni(topology: Box) -> None:
     if transit_vni == vni_start:                                # If we had to assign the default value, increment the default transit VNI
       vni_start = get_next_vni(vni_start,vni_list)
 
+  for vrf_name,vrf_data in topology.vrfs.items():               # Third pass: set shared VNI values across VRFs
+    if vrf_data is None:                                        # Skip empty VRF definitions
```
Comment (jbemmel) [inline netsim/modules/evpn.py:122]: Should not happen

**Comment by ipspace** [comment]
Thanks for the review! The `vrf is None` case happens when you declare **vrfs** without the **vrf** module. Obvious **netlab** reports an error and exist, but the global attributes are checked after **pre_transform** (because they might be adjusted/added during the pre_transform), so unfortunately it's perfectly possible to have **vrfs** with empty VRFs in an invalid topology (and yet we netlab shouldn't crash).

**Review by jbemmel — CHANGES_REQUESTED** [review]
I agree this is a superior solution, I like it

Few dots to be corrected - learned that Box will happily assign vrf.evpn_transit_vni=x into vrf -> evpn -> transit_vni=x, and vrf.evpn_bull=crap into vrf -> evpn_bull = crap

File: docs/module/evpn.md [hunk 3]
```diff
@@ -89,6 +89,13 @@ The default value of VRF EVPN Instance identifier is the VLAN ID of the first VL
...
 * Asymmetric IRB requires no extra parameters[^NS]
-* Symmetric IRB needs a transit VNI that has to be set with the **evpn.transit_vni** parameter. This parameter could be set to an integer value or to *True* in which case the EVPN configuration module auto-assigns a VNI to the VRF. Note that the EVI value used in this case is currently based on the VRF ID (vrfidx)
+* Symmetric IRB needs a transit VNI that has to be set with the **evpn.transit_vni** parameter.
+* You can set the EVI value with **evpn.evi** parameter.
+
+The **evpn.transit_vni** parameter must specify a globally unique VNI value. It could be set to:
+
+* *True*: EVPN configuration module auto-assigns a unique VNI to the VRF.
+* An *integer value*: static VNI assignment, checked for uniqueness
+* Name of *another VRF*: the **evpn_transit_vni** value is copied from that VRF. Use this setting for complex topologies where VRFs with different connectivity requirements have to share the transit VXLAN segment.
```
[the remark this answers is not in the document]
Comment (ipspace) [reply docs/module/evpn.md:99]: Thanks a million for all the dots (and that Box behavior is beyond imaginable). Noticed this bit a bit too late, so I amended the commit.

File: netsim/modules/evpn.py [hunk 4]
```diff
@@ -116,6 +118,27 @@ def vrf_transit_vni(topology: Box) -> None:
...
+        'evpn')
+      continue
+    foreign_vni = data.get_from_box(topology.vrfs,f'{transit_vni}.evpn.transit_vni')
+    if not data.is_true_int(foreign_vni):
+      common.error(
+        f'evpn.transit_vni "{transit_vni}" in VRF {vrf_name} refers to a VRF that does not have a valid evpn.transit_vni',
+        common.IncorrectValue,
+        'evpn')
+      continue
+    vrf_data.evpn_transit_vni = foreign_vni
```
[the remark this answers is not in the document]
Comment (ipspace) [reply netsim/modules/evpn.py:140]: Fixed & amended the commit.

**Comment by jbemmel** [comment]
We could consider to change things such that the vrf module gets invoked whenever a topology has a 'vrfs' section - regardless if the user remembered to activate it as a module.

That way, the responsibility of sanitizing that section can remain squarely with the vrf module, and other code that references vrfs doesn't have to remember to double-check the sanity of the user input. Same applies for vrf naming for example.

Similar consideration for the vlan module

**Pull request merged**